models/vrp.rb

```diff
@@ -140,10 +140,94 @@ def self.create(hash, delete = true)
       vrp
     end
 
+    def self.convert_shipments_to_services(hash)
+      hash[:services] ||= []
+      hash[:relations] ||= []
+      service_ids = hash[:services].map{ |service| service[:id] }
+      hash[:shipments]&.each{ |shipment|
+        linked_ids = []
+        %i[pickup delivery].each{ |part|
+          service = Oj.load(Oj.dump(shipment))
+          service[:original_id] = shipment[:id]
+          service[:id] += "_#{part}"
+          service[:id] += '_conv' while service_ids.any?{ |id| id == service[:id] } # protect against id clash
+          linked_ids << service[:id]
+          service[:type] = part.to_sym
+          service[:activity] = service[part]
+          %i[pickup delivery direct maximum_inroute_duration].each{ |key| service.delete(key) }
+
+          if part == :delivery
+            service[:quantities]&.each{ |quantity|
+              quantity[:value] = -quantity[:value] if quantity[:value]
+              quantity[:setup_value] = -quantity[:setup_value] if quantity[:setup_value]
+            }
+          end
+          hash[:services] << service
+        }
+        convert_relations_of_shipment_to_services(hash, shipment[:id], linked_ids[0], linked_ids[1])
+        hash[:relations] << { type: :shipment, linked_ids: linked_ids }
+        hash[:relations] << { type: :sequence, linked_ids: linked_ids } if shipment[:direct]
+        max_lapse = shipment[:maximum_inroute_duration]
+        hash[:relations] << { type: :maximum_duration_lapse, linked_ids: linked_ids, lapse: max_lapse } if max_lapse
+      }
+      convert_shipment_within_routes(hash)
+      hash.delete(:shipments)
+    end
+
+    def self.convert_relations_of_shipment_to_services(hash, shipment_id, pickup_service_id, delivery_service_id)
+      hash[:relations].each{ |relation|
+        case relation[:type]
+        when :minimum_duration_lapse, :maximum_duration_lapse
+          relation[:linked_ids][0] = delivery_service_id if relation[:linked_ids][0] == shipment_id
+          relation[:linked_ids][1] = pickup_service_id if relation[:linked_ids][1] == shipment_id
+          relation[:lapse] ||= 0
+        when :same_route
+          relation[:linked_ids].each_with_index{ |id, id_i|
+            next unless id == shipment_id
+
+            relation[:linked_ids][id_i] = pickup_service_id # which will be in same_route as delivery
+          }
+        when :sequence, :order
+          if relation[:linked_ids].any?{ |id| id == shipment_id }
+            msg = 'Relation between shipment pickup and delivery should be explicitly specified for `:sequence` and `:order` relations.'
+            raise OptimizerWrapper::DiscordantProblemError.new(msg)
+          end
+        else
```

Comment:
Could we use elsif to avoid a new indentation level ?

Reply:
The `else` comes from the `case` not from a `if`. And this generates a syntax error :/